nemo_skills/code_execution/local_sandbox/local_sandbox_server.py

```diff
@@ -13,41 +13,95 @@
+
+        project_path = "/lean4/my_project"
+        with tempfile.NamedTemporaryFile(dir=project_path, delete=False, suffix=".lean") as temp_file:
+            temp_file_name = temp_file.name
+            temp_file.write(generated_code.encode('utf-8'))
+
+        result = subprocess.run(
+            ['lake', 'env', '--dir', project_path, 'lean', temp_file_name],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            timeout=timeout,
+            cwd=project_path,  # Ensure we are in the correct working directory
+        )
+
+        if result.returncode == 0:
+            process_status = "completed"
+        else:
+            process_status = "failed"
+
+        return {
+            "process_status": process_status,
+            "stdout": result.stdout.decode('utf-8'),
+            "stderr": result.stderr.decode('utf-8'),
+        }
+
+    except subprocess.TimeoutExpired:
+        return {"process_status": "timeout", "stdout": "Timed out", "stderr": "Timed out"}
+    except Exception as e:
+        print(f"Error: {str(e)}")
+        return {"process_status": "error", "stdout": "", "stderr": str(e)}
+    finally:
+        # Safely remove the temporary file if it was created
+        if temp_file_name and os.path.exists(temp_file_name):
+            os.remove(temp_file_name)
+
+
 def execute_code_subprocess(generated_code, queue):
-    # this can be overriden inside generated code, so it's not a guaranteed protection
-    limit = 1024 * 1024 * 1024 * 10  # 10gb - somehow with a smaller limit the server dies when numpy is used
+    limit = 1024 * 1024 * 1024 * 10  # 10 GB memory limit
     resource.setrlimit(resource.RLIMIT_AS, (limit, limit))
     resource.setrlimit(resource.RLIMIT_DATA, (limit, limit))
     resource.setrlimit(resource.RLIMIT_STACK, (limit, limit))
 
     sys.stdout = StringIO()
-    exec(generated_code, {})
-    queue.put(sys.stdout.getvalue())
+    try:
+        exec(generated_code, {})
+        queue.put(sys.stdout.getvalue())
+    except Exception as e:
+        queue.put({"process_status": "error", "stdout": "", "stderr": str(e)})
 
 
+# Main Flask endpoint to handle execution requests
 @app.route("/execute", methods=["POST"])
 def execute():
     generated_code = request.json['generated_code']
     timeout = request.json['timeout']
-    # running in a separate process to ensure any kind of crashes are properly handled
-    queue = multiprocessing.Queue()
-    process = multiprocessing.Process(target=execute_code_subprocess, args=(generated_code, queue))
-    process.start()
-    process.join(timeout=timeout)
-    if process.is_alive():  # didn't finish successfully
-        process.kill()
-        return '{"process_status": "timeout", "stdout": "Timed out", "stderr": "Timed out"}'
-    return queue.get()
+    language = request.json.get('language', 'python')
+
+    if language == 'python':
+        return execute_python(generated_code, timeout)
+    elif language == 'lean4':
```

Comment:
I suggest to just call it "lean", since for python we don't specify the version

Reply:
lean4 is significantly different from lean3 mainly in syntax. Lean3 code cannot be compiled in lean4. Even the official github repos use lean4 as their name 
https://github.com/leanprover/lean4
As well as they use mathlib4 instead of just mathlib
https://github.com/leanprover-community/mathlib4

That's why I put lean4 instead of just lean

Reply:
Well, the same can be said for python, but we don't really need to call it python3, since no one is using python2 anymore. But anyway, I don't have a strong opinion about that, so we can keep it called lean4 if you think it's clearer